### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by bennett000**
The tests past and the example mostly runs but this is a little premature for a PR, and it needs some work:

- the example's '0' does not increment
- the dependencies may have been changed too aggressively and might break, I `npm prune`d before testing, but the world is an imperfect place...
- the git history is noisy

## Updates

- ported the library to TypeScript
- migrated example and core from `tsd` to `typings`
- pointed to Redux typings that support generics.  Unfortunately I did not get the official Redux typings to work, I got a slew of compile errors when I pointed to the bleeding edge
- Updated Angular 2 to beta 8
- Changed example to use life cycle hooks for binding, instead of constructor

_update_ I should have realized there was Travis, that will point me to the dependency errors I fear!

### CODE REVIEW

**Comment by wbuchwalter** [comment]
Nice work! :+1: 
I will take a look at that later today.

**Comment by bennett000** [comment]
Thanks @wbuchwalter.  I heard you're busy with a new job, congrats.

I know @kosz had some issues with my earlier version, hopefully he's able to get this PR working on his system. I'd also like to fix the counter view in the example as it's not updating.  Plus it would be great to use to official Redux bindings, I did not look too closely at all the `tsc` errors they gave me.

File: examples/counter/typings/browser/ambient/es6-shim/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,670 @@
+// Generated by typings
```
Comment (kosz) [inline examples/counter/typings/browser/ambient/es6-shim/index.d.ts]: @bennett000 should the typing files be merged in for the examples ?

File: tmp/tsreq/src/components/utils/shallowEqual.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+function shallowEqual(objA, objB) {
```
Comment (kosz) [inline tmp/tsreq/src/components/utils/shallowEqual.js]: should this `tmp` folder be merged in ?

**Comment by bennett000** [comment]
Thanks @kosz, the typings should _not_ be merged in, I'll check on tmp

File: package.json [hunk 3]
```diff
@@ -4,9 +4,12 @@
   "description": "Angular 2 bindings for Redux",
   "main": "./lib/index.js",
   "scripts": {
-    "build": "rm -rf lib && `npm bin`/babel src --out-dir lib",
-    "test": "mocha --compilers js:babel/register --recursive"
+    "build": "rm -rf lib; tsc",
+    "postinstall": "npm run typings",
+    "test": "mocha --compilers js:babel/register --recursive --require ts-node/register",
+    "typings": "typings install"
```
Comment (kosz) [inline package.json]: Would this fail if `typings` is not installed globally ?

**Comment by bennett000** [comment]
Transpiled output has been included in the latest push, this was to test the TypeScript version through `npm install` on other projects.  Before any merging the last commit needs to be rebased out, because keeping transpiled code is sub-optimal.

**Comment by e-schultz** [comment]
@bennett000 could you re-open this PR against the typescript branch instead of master - then we can merge it into there and continue work on it until it's ready for a full release

**Comment by bennett000** [comment]
This PR is going to be closed and re-opened against the typescript branch

**Pull request closed without merging**